feat(docs): add listening to events and eventBus (#35)
bpmn-io/docs.bpmn.io: docs/Modules/common-modules/event-bus.md

It is used to communicate across a diagram instance.
Several components of a diagram can use it to listen to and to broadcast events.

It provides the following methods:

* [on](#on)
* [once](#once)
* [off](#off)
* [createEvent](#createEvent)
* [fire](#fire)
* [handleError](#handleError)

## on

### Usage

on(events: _String_ or _Array< string >_, [priority=1000]: _number_, callback: _Function_, [that]:_Object_): _undefined_

### Parameters

* `events`: name of the event or _Array_ of event names.
* `priority`: optional parameter specifying the priority in which this listener is called, where a higher number represents a higher priority and therefore leads to earlier being called. If not provided, default of `1000` will be used.
* `callback`: function which will be invoked with `event, ...additionalArguments`. Returning anything but `undefined` from this will stop the listener propagation. Returning `false` will prevent the execution of the default action for that event.
*  `that`: optional parameter which can be used to pass context (`this`) to the callback.

### Return Value

_undefined_

### Code Example
```javascript
eventBus.on('saveSVG.start', 1500, function(event) {

>   // access event type

Reviewer: Same as above, let's add a new line above.

  // [...]
});
```

## once

### Usage

once(events: _String_ or _Array< string >_, [priority=1000]: _number_, callback: _Function_, [that]:_Object_): _undefined_

### Parameters

* `events`: name of the event or _Array_ of event names.
* `priority`: optional parameter specifying the priority in which this listener is called, where a higher number represents a higher priority and therefore leads to earlier being called. If not provided, default of `1000` will be used.
* `callback`: function which will be invoked with `event, ...additionalArguments`. Returning anything but `undefined` from this will stop the listener propagation. Returning `false` will prevent the execution of the default action for that event.
*  `that`: optional parameter which can be used to pass context (`this`) to the callback.

### Return Value

_undefined_

## off

### Usage

off(events: _String_ or _Array< string >_, [callback]: _Function_): _undefined_

### Parameters

* `events`: name of the event or _Array_ of event names.
* `callback`: optional parameter specifying the callback listener to be removed.

### Return Value

_undefined_

### Code Example
```javascript
function listener(event) {

>   // access event type

Reviewer: Same as above, let's add a new line above.

  event.type; // 'saveSVG.start'

  // stop propagation to other listeners
  event.stopPropagation();

  // prevent event default
  event.preventDefault();
}

eventBus.on('saveSVG.start', 1500, listener);
eventBus.off('saveSVG.start', listener);
```

## createEvent

### Usage

createEvent(data: _Object_): _Object_

### Parameters

* `data`: the event data.

### Return Value

An _InternalEvent_ object which can be passed to [fire](#fire).

## fire

### Usage

fire(name: _String_, [data]: _Object_, [arguments]: _...Object_): _Object_

> ### Parameters

Reviewer: we should make it consistent whether to use an article ("the"), e.g.

> the event name.

Or leave it without, e.g.

> optional additional arguments ...

Author: I agree, let's have a separate discussion about this covering all content that we have so far. Here and in general I would go for the simpler alternative without the article

* `name`: the event name.
* `data`: the optional event object. If this is not already an instance of `InternalEvent`, the `fire` method will internal wrap it as such (using [`createEvent`](createEvent)).
* `arguments`: optional additional arguments to be passed to the callback functions.

### Return Value

* `false` if the default action was prevented by listeners.
* `undefined` if no listeners were registered for that event.
* _Object_ if the invoked listener with the highest priority returns it.

### Code Example

> // TODO: how many examples / how complex examples?

Reviewer: I'd add one rather small example and one complex one.

For the complex example, maybe we can use listener chaining, as we do it in the [`HoverFix`](https://github.com/bpmn-io/diagram-js/blob/develop/lib/features/dragging/HoverFix.js#L70) (btw. one of my all-time favorites!)

```js
var count = 0;

function countListener() {
  
  // stop the counter once 'break' was fired   
  eventBus.once('break', function() {
    eventBus.off('tick', countListener);

    console.log('count', count);
    count = 0;
  });

  // count how often event was fired after first execution
  count++;
};

eventBus.on('tick', countListener);

Author: Great example indeed! Thanks

grand_parent: "Modules"
---

# eventBus

A general purpose event bus.

> It is used to communicate across a diagram instance.

Reviewer 2: We need to clarify the terminology. We shouldn't use module, service, and component for the same thing. My proposal would be _module_ since that is what they actually are. This could then be changed to:

_It is used to communicate between modules of a diagram instance without coupling them._

Author: I like that :+1:

Several components of a diagram can use it to listen to and to broadcast events.

It provides the following methods:

* [on](#on)
* [once](#once)
* [off](#off)
* [createEvent](#createEvent)
* [fire](#fire)
* [handleError](#handleError)

## on

### Usage

on(events: _String_ or _Array< string >_, [priority=1000]: _number_, callback: _Function_, [that]:_Object_): _undefined_

### Parameters

* `events`: name of the event or _Array_ of event names.

> * `priority`: optional parameter specifying the priority in which this listener is called, where a higher number represents a higher priority and therefore leads to earlier being called. If not provided, default of `1000` will be used.

Reviewer 2: ```suggestion
* `priority`: optional parameter specifying the priority with which the listener is called, where a higher number represents a higher priority and therefore leads to the listener being called earlier. If not provided, default of `1000` will be used.
```

_undefined_

### Code Example

```javascript
eventBus.on('saveSVG.start', 1500, function(event) {

  // access event type
  event.type; // 'saveSVG.start'

  // stop propagation to other listeners
  event.stopPropagation();

  // prevent event default
  event.preventDefault();

  // Some custom logic
  // [...]
});
```

## once

### Usage

once(events: _String_ or _Array< string >_, [priority=1000]: _number_, callback: _Function_, [that]:_Object_): _undefined_

### Parameters

* `events`: name of the event or _Array_ of event names.
* `priority`: optional parameter specifying the priority in which this listener is called, where a higher number represents a higher priority and therefore leads to earlier being called. If not provided, default of `1000` will be used.
* `callback`: function which will be invoked with `event, ...additionalArguments`. Returning anything but `undefined` from this will stop the listener propagation. Returning `false` will prevent the execution of the default action for that event.
*  `that`: optional parameter which can be used to pass context (`this`) to the callback.

### Return Value

_undefined_

### Code Example

```javascript
let count = 0;

> function countListener() {

Reviewer 2: This example is quite hard to understand and I'm not sure how it's supposed to work.

  // stop the counter once 'break' was fired   
  eventBus.once('break', function() {

    // Don't tick anymore
    eventBus.off('tick', countListener);

    console.log('count', count);
    count = 0;
  });

  // count how often event was fired after first execution
  count++;
};

eventBus.on('tick', countListener);
```

## off

### Usage

off(events: _String_ or _Array< string >_, [callback]: _Function_): _undefined_

### Parameters

* `events`: name of the event or _Array_ of event names.
* `callback`: optional parameter specifying the callback listener to be removed.

### Return Value

_undefined_

### Code Example

> ```javascript

Reviewer 2: This example doesn't really help. I'd be more interested in the actual effect of calling `off`.

eventBus.fire('foo'); // 'Foo' 'Bar'

eventBus.off('foo', fooListener);

eventBus.fire('foo'); // 'Bar'

eventBus.off('foo');

eventBus.fire('foo'); // ''
```

## createEvent

### Usage

createEvent(data: _Object_): _Object_

### Parameters

* `data`: the event data.

### Return Value

An _InternalEvent_ object which can be passed to [fire](#fire).

## fire

### Usage

fire(name: _String_, [data]: _Object_, [arguments]: _...Object_): _Object_

### Parameters

* `name`: event name.
* `data`: optional _InternalEvent_ object. If this is not already an instance of `InternalEvent`, the `fire` method will internal wrap it as such (using [`createEvent`](createEvent)).
* `arguments`: optional additional arguments to be passed to the callback functions.

### Return Value

* `false` if the default action was prevented by listeners.
* `undefined` if no listeners were registered for that event.
* _Object_ if the invoked listener with the highest priority returns a value.

### Code Example

```javascript

> function listener(...args) {

Reviewer 2: Maybe we should promote the way we use the event bus internally in this example. If the first parameter is an object it will be assigned to the event itself which allows you to do

```javascript
eventBus.on('foo', ({ answer }) => console.log(answer));

eventBus.fire('foo', { answer: 42 });
```

At least this is the way we do it everywhere in our toolkit.

Author: Good point. I would still like to keep the existing example, since then both examples represent basically two alternatives of how to use `fire`

* `callback`: function which will be invoked with `event, ...additionalArguments`. Returning anything but `undefined` from this will stop the listener propagation. Returning `false` will prevent the execution of the default action for that event.
*  `that`: optional parameter which can be used to pass context (`this`) to the callback.

### Return Value

_undefined_

### Code Example

```javascript
function listener(event) {

  // Perform some operation you only want to do once
  // [...]

  return 'One-time operation finished';
}

eventBus.once('onceEvent', listener);

const firstEventFired = eventBus.fire('onceEvent'),
      secondEventFired = eventBus.fire('onceEvent');

console.log(firstEventFired); // 'One-time operation finished'
console.log(secondEventFired); // undefined
```

## off

### Usage

off(events: _String_ or _Array< string >_, [callback]: _Function_): _undefined_

### Parameters

* `events`: name of the event or _Array_ of event names.
* `callback`: optional parameter specifying the callback listener to be removed.

### Return Value

_undefined_

### Code Example

```javascript

> function listener1(event) {

Reviewer 2: I'd go for an even simple example that doesn't also show priorities and return values.

Something like this:

```javascript
function fooListener(event) {
  console.log('Foo');
}

function barListener(event) {
  console.log('Bar');
}

eventBus.on('foo', fooListener);
eventBus.on('foo', barListener);

eventBus.fire('foo'); // 'Foo Bar'

eventBus.off('foo', fooListener);

eventBus.fire('foo'); // 'Bar'

eventBus.off('foo');

eventBus.fire('foo'); // ''
```

Author: Good point
